src/ipa_pc/mod.rs

```diff
@@ -672,21 +725,45 @@ impl<G: AffineCurve, D: Digest> PolynomialCommitment<G::ScalarField> for InnerPr
             let hiding_bound = labeled_polynomial.hiding_bound();
             let commitment = labeled_commitment.commitment();
 
-            combined_polynomial += (cur_challenge, polynomial);
-            combined_commitment_proj += &commitment.comm.mul(cur_challenge);
+            let p_len = polynomial.coeffs.len();
+
+            if p_len > key_len {
+
+                let mut polynomial_lc = Polynomial::zero();
+                for i in 0..p_len / key_len + if p_len % key_len != 0 { 1 } else { 0 } {
+                    let is = i * key_len;
+                    let poly_single = Polynomial::from_coefficients_slice(&polynomial.coeffs[i * key_len..core::cmp::min((i + 1) * key_len, p_len)]);
+                    polynomial_lc += (point.pow(&[is as u64]), &poly_single);
+                }
+                combined_polynomial += (cur_challenge, &polynomial_lc);
+
+                let mut comm_lc = <G::Projective as ProjectiveCurve>::zero();
+                for (i, comm_single) in commitment.comm.iter().enumerate() {
+                    let is = i * key_len;
+                    comm_lc += &comm_single.mul(point.pow(&[is as u64]));
+                }
+                combined_commitment_proj += &comm_lc.mul(&cur_challenge);
+
+            } else {
+
+                combined_polynomial += (cur_challenge, polynomial);
+                combined_commitment_proj += &commitment.comm[0].mul(cur_challenge);
+            }
 
             if hiding_bound.is_some() {
                 has_hiding = true;
-                combined_rand += &(cur_challenge * &randomness.rand);
+                combined_rand += &(cur_challenge * &randomness.randomness().rand);
```

Comment:
As we will have individual blinding randomnesses for each of the segment polynomials, we will have to take their linear combination (using the powers of the query point) instead.

Reply:
Done